File: leetcode/2331.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

class Solution {
    dfs(curr) {
        if (curr === null) return;
        
        this.dfs(curr.left);
        this.dfs(curr.right);
        
        if (curr.val === 2 || curr.val === 3) {
            if (curr.left !== null && curr.right !== null) {
                const y = curr.right.val;
                const x = curr.left.val;
                
                if (curr.val === 3)
                    curr.val = x && y;
                else
                    curr.val = x || y;
            }
        }
    }
    
    evaluateTree(root) {
        if (root === null)
            return false;

        this.dfs(root);
        return root.val;
    }
}

File: leetcode/2331.cpp
// Runtime 12 ms Beats 46.64% of users with C++
// Memory 17.20 MB Beats 100.00% of users with C++

// Time , Memory Complexity : O(N), O(N)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* curr){
        if(curr == nullptr) return;
        
        dfs(curr->left);
        dfs(curr->right);
        
        if(curr->val == 2 || curr->val == 3){
            if(curr->left != nullptr && curr->right != nullptr){
                bool y = curr->right->val;
                bool x = curr->left->val;
                
                if(curr->val == 3)
                    curr->val = x && y;
                else
                    curr->val = x || y;
            }
        }
    }
    
    bool evaluateTree(TreeNode* root) {
        if (root == nullptr)
            return false;

        dfs(root);
        return root->val;
    }
};

